agent/proxycfg/state.go

```diff
@@ -1048,8 +1048,8 @@ func (s *state) handleUpdateIngressGateway(u cache.UpdateEvent, snap *ConfigSnap
 			return fmt.Errorf("invalid type for response: %T", u.Result)
 		}
 
-		var upstreams structs.Upstreams
 		watchedSvcs := make(map[string]struct{})
+		upstreamsMap := make(map[IngressListenerKey]structs.Upstreams)
```

Comment:
minor optimization, you could set the initial capacity of the map with:

```go
upstreamMap := make(map[IngressListenerKey]structs.Upstreams, len(services.Services))
```

to avoid additional allocations as the map grows. Probably not a big deal in this case unless  there are a very large number of services.

Reply:
The only reason I hesitate on this one is because I think the `len(services.Services)` will be much greater than the number of unique `IngressListenerKey`s. That is to say, I think most ingress gateways will have a small number of ports behind which a large number of services are exposed.

Reading golang's `make` documentation, it seems like specifying a size for a map indicates you expect that many key/value pairs.

Reply:
Ah yes, exactly. I think you are right, in this case it probably is not a good change, I was thinking the id would be unique per service. :+1: